### ISSUE TITLE
Web UI single sign-on against OIDC provider

### ISSUE THREAD

**Issue by marians** [giantswarm/roadmap#218]
## User story

- _As a cluster admin, I want to log in to the Giant Swarm web UI using the same identity I use for authenticating with the Control Plane Kubernetes API._

## Background

The Giant Swarm web UI (a.k.a. [happa](https://github.com/giantswarm/happa)) currently requires user authentication against the [Giant Swarm REST API](https://docs.giantswarm.io/basics/api/#rest-api). User identities here are individual for each installation.

As of October 2020, many – but not all – customers also have access to the [Management API](https://docs.giantswarm.io/ui-api/management-api/). Authentication for this API is done via OIDC/[dex](https://github.com/dexidp/dex) using the customer's own identity providers.

Giant Swarm decided to phase out the REST API eventually, making the Management API the primary interface for clients. This means that the web UI (happa) has to transition from using the REST API to using the Management API, too. As we want to allow for this transition to take some time, we need a way to let happa communicate with both APIs at the same time. However we still want to have happa users authenticate using only one identity. (The alternative, letting users handle several identities and authenticate two times, would likely lead to a very bad user experience.)

## Solution

- All users authenticate with happa using OIDC.
- happa sends the same JWT token to the REST API and to the Management API.
- The REST API accepts JWT tokens (given claims match the expectations) and impersonates the user issuing the incoming request.
- Organization membership will be based on RBAC roles and bindings in Kubernetes.
- In happa, we handle cases where permissions are missing (e. g. only read, no write access) specifically.

## Definition of Done

User logs in to happa using OIDC account, doing everything she could do with the REST API account.

## Dependencies

- All customers need Management API access authenticated via dex/OIDC.
- REST API must be able to figure out org membership for an OIDC user. Customer admins need to be able to manage organizations and assign users/groups to organizations.

**Comment by snizhana-dynnyk** [comment]
Hi @giantswarm/team-batman :) 
This issue is the last thing that is blocking Happa migration to CP k8s API, once this is done other teams can start migrating their functionality to Happa.
It is currently in team Biscuit but we don’t feel comfortable working on it, we don’t have FE engineers and a little experience implementing middleware (API).

So asking you if you can take over driving this since @oponder has the best understanding of what needs to be done. Of course you can count on all needed support from Biscuit in everything related to organizations mapping etc.

**Comment by cokiengchiara** [comment]
@oponder since this will unblock other teams, WDYT let's discuss prioritizing this after you finish current WIP? Will raise in our 1 on 1 on Thursday. But feel free to ping earlier if you want.

**Comment by snizhana-dynnyk** [comment]
Thank you! Also just make it clear - it is not like other teams can't wait for it to be done to start migration :) So it is not super urgent, this is just the next thing that needs to be done to proceed with Happa migration. It is totally up to you how to prioritise, we just think this makes more sense in your team <3

**Comment by cokiengchiara** [comment]
Discussed w @oponder today. He's happy to leads this. Will do it after https://github.com/giantswarm/giantswarm/issues/13652 😀

**Comment by JosephSalisbury** [comment]
me / vol / oponder chatted:

- we reckon it's straightforward for us to precreate clusterroles for customer usage that have various styles of access (e.g: creating all clusters, listing clusters, managing apps), and we need to create a delightful experience here (e.g: kgs / happa setup)
    - need documentation for this
- leaving very fancy permission management out of scope for now
- biscuit doesn't do anything right now due to hive sprint, it doesn't seem that we're blocking
- oponder thinks about joining hive sprint
- possibly more straightforward to avoid 'mashing' GS API and MC API, and just go straight to MC API
- tl;dr - this is complicated, it possibly changes a lot w/ hive sprint, let's do stuff there

---

<details>
<summary>joe's real crazy flow of conciousness notes:</summary>
there is no current mapping between companyd organisations and organisation CRs, so using JWT tokens in the old giantswarm api doesn't make sense

on azure, cluster CRs are in organisation namespaces. on aws / kvm, cluster CRs are in default namespace

oponder: phase out GS API, talk against only MC API. adrian has a bridge. how do we do organisations? lives inside companyd (joint able between companyd and userd). oidc token (from dex) can talk to MC API, but doesn't have same idea as limiting resources. what are we doing w/ organisations in MC API?

vol: organisations don't have anything related to organisation scope / allowed access - organisation CR manages namespaces / rbac to allow access to them. how does rbac fit into happa story? for user access to MC API, we map IDP groups w/ clusterroles in MC. to do same for happa, when we configure organisation, organisation should map to group(s) in IDP.

vol: giant swarm employees have mapping from github group to cluster admin role. when we create, we ask customers which IDP group to map to not cluster admin role, but similar

oponder: how to list which organisations i belong to? 

- map from IDP groups to organisation CRs

workflow (e.g: user w/ token wants to list cluster CRs):
1. log into happa w/ third party identity provider (e.g: github)
2. get token from github, has information about IDP groups
3. organisation CR: acme -> namespace: org-acme, cluster CRs live in that namespace

organisation:
[]idp group
rbac-operator: shitty CR -> cluster 

happa startup (wants to get list of all organisations)
- can i list all organisation CRs? does that get filtered? if you get no organisations / an error message for permissions -> that's fine
- can current admins list organisations?

log in via idp 
you're in some group
that is bound to some pre-created clusterrole
</details>

**Comment by marians** [comment]
ℹ️  Updated the main comment to use our current terminology and fill in some blanks.

**Comment by axbarsan** [comment]
Customer SSO is enabled by default on testing installations. It will be enabled on customer installations once we get all the permission things out of the way.

GS Staff SSO (aka admin login) is now using `GitHub` instead of `Auth0` ✅

**Comment by cokiengchiara** [comment]
* Is it worth announcing? I think not to blog, but to customers?
* Update docs?

**Comment by cokiengchiara** [comment]
Communications are being taken care of here https://github.com/giantswarm/giantswarm/issues/16042
